Mathlib/Mathport/Notation.lean

```diff
@@ -268,46 +265,42 @@ where
-      let ppTypes ← getPPOption getPPPiBinderTypes -- the same option controlling ∀
-      let dom ← withBindingDomain delab
-      withBindingBodyUnusedName <| fun x => do
-        let x : Ident := ⟨x⟩
-        let binder ←
-          if prop && !isDep then
-            -- this underscore is used to support binder predicates, since it indicates
-            -- the variable is unused and this binder is safe to merge into another
-            `(extBinderParenthesized|(_ : $dom))
-          else if prop || ppTypes then
-            `(extBinderParenthesized|($x:ident : $dom))
-          else
-            `(extBinderParenthesized|($x:ident))
-        -- Now use the body of the lambda for `lit` for the next iteration
-        let s ← s.captureSubexpr lit
-        let s ← s.pushBinder binder
-        matchScoped lit scopeId smatcher s
-  catch _ =>
-    return s
-
-/-- Like `matchScoped` but ensures that it matches at least one binder. -/
-partial def matchScoped' (lit scopeId : Name) (smatcher : Matcher) : Matcher := fun s => do
-  guard <| s.scopeState.isEmpty
-  let s ← matchScoped lit scopeId smatcher s
-  guard <| !s.scopeState.isEmpty
-  return s
+(which must be a lambda) to produce a binder, and loops. -/
+partial def matchScoped (lit scopeId : Name) (smatcher : Matcher) : Matcher := go #[] where
+  go (binders : Array (TSyntax ``extBinderParenthesized)) : Matcher := fun s => do
+    -- `lit` is bound to the SubExpr that the `scoped` syntax produced
+    s.withVar lit do
+    try
+      -- Run `smatcher` at `lit`, clearing the `scopeId` variable so that it can get a fresh value
+      let s ← smatcher {s with vars := s.vars.erase scopeId}
+      s.withVar scopeId do
+        guard (← getExpr).isLambda
+        let prop ← try Meta.isProp (← getExpr).bindingDomain! catch _ => pure false
+        let isDep := (← getExpr).bindingBody!.hasLooseBVar 0
+        let ppTypes ← getPPOption getPPPiBinderTypes -- the same option controlling ∀
+        let dom ← withBindingDomain delab
+        withBindingBodyUnusedName <| fun x => do
+          let x : Ident := ⟨x⟩
+          let binder ←
+            if prop && !isDep then
+              -- this underscore is used to support binder predicates, since it indicates
+              -- the variable is unused and this binder is safe to merge into another
+              `(extBinderParenthesized|(_ : $dom))
+            else if prop || ppTypes then
+              `(extBinderParenthesized|($x:ident : $dom))
+            else
+              `(extBinderParenthesized|($x:ident))
+          -- Now use the body of the lambda for `lit` for the next iteration
+          let s ← s.captureSubexpr lit
+          -- TODO merge binders as an inverse to `satisfies_binder_pred%`
+          let binders := binders.push binder
+          go binders s
+    catch _ =>
+      guard <| !binders.isEmpty
+      if let some _binders := s.scopeState then
+        return s -- TODO: check for equality?
```

Comment:
I would really rather that we did the equality check, since without it you could get some very misleading output.

Reply:
the trouble is that we have syntax here, not exprs. We could compare the syntax but that's likely going to have some weird effects of its own.